File: go/problems/unique_number_of_occurances.go
package problems

// https://leetcode.com/problems/unique-number-of-occurrences/
func UniqueOccurrences(arr []int) bool {
	frequencyMap := make(map[int]int)
	for _, item := range arr {
		if _, ok := frequencyMap[item]; ok {
			frequencyMap[item] += 1
		} else {
			frequencyMap[item] = 1
		}
	}
	peek := make(map[int]bool)
	for _, value := range frequencyMap {
		if _, ok := peek[value]; ok {
			peek[value] = true
			return false
		}
	}
	return true
}

File: rust/src/problems/unique_number_of_occurances.rs
// https://leetcode.com/problems/unique-number-of-occurrences/
use std::collections::{HashMap, HashSet};

pub fn unique_occurrences(arr: Vec<i32>) -> bool {
  let frequency: HashMap<i32, i32> = arr.iter().fold(HashMap::new(), |mut acc, curr| {
    *acc.entry(*curr).or_insert(0) += 1;
    return acc;
  });
  let frequency_hash_set: HashSet<i32> = frequency.values().cloned().collect();
  return if frequency.values().len() == frequency_hash_set.len() {
    true
  } else {
    false
  };
}
